core/app/models/spree/tax_rate.rb

```diff
@@ -29,13 +29,56 @@ class TaxRate < Spree::Base
 
     scope :by_zone, ->(zone) { where(zone_id: zone) }
 
-    # Gets the array of TaxRates appropriate for the specified order
+    # Finds geographically matching tax rates for an order's tax zone.
+    # We do not know if they are/aren't applicable until we attempt to apply these rates to
+    # the items contained within the Order itself.
+    # For instance, if a rate passes the criteria outlined in this method,
+    # but then has a tax category that doesn't match against any of the line items
+    # inside of the order, then that tax rate will not be applicable to anything.
+    # For instance:
+    #
+    # Zones:
+    #   - Spain (default tax zone)
+    #   - France
+    #
+    # Tax rates: (note: amounts below do not actually reflect real VAT rates)
+    #   21% inclusive - "Clothing" - Spain
+    #   18% inclusive - "Clothing" - France
+    #   10% inclusive - "Food" - Spain
+    #   8% inclusive - "Food" - France
+    #   5% inclusive - "Hotels" - Spain
+    #   2% inclusive - "Hotels" - France
+    #
+    # Order has:
+    #   Line Item #1 - Tax Category: Clothing
+    #   Line Item #2 - Tax Category: Food
+    #
+    # Tax rates that should be selected:
+    #
+    #  21% inclusive - "Clothing" - Spain
+    #  10% inclusive - "Food" - Spain
+    #
+    # If the order's address changes to one in France, then the tax will be recalculated:
+    #
+    #  18% inclusive - "Clothing" - France
+    #  8% inclusive - "Food" - France
+    #
+    # Note here that the "Hotels" tax rates will not be used at all.
+    # This is because there are no items which have the tax category of "Hotels".
+    #
+    # Under no circumstances should negative adjustments be applied for the Spanish tax rates.
+    #
+    # Those rates should never come into play at all and only the French rates should apply.
     def self.match(order_tax_zone)
       return [] unless order_tax_zone
-      rates = includes(zone: { zone_members: :zoneable }).load.select do |rate|
-        # Why "potentially"?
-        # Go see the documentation for that method.
-        rate.potentially_applicable?(order_tax_zone)
+      all_rates = includes(zone: { zone_members: :zoneable }).load
+
+      rates_for_order_zone = []
+      rates_for_default_zone = []
+
+      all_rates.each do |rate|
+        rates_for_order_zone << rate if rate.zone.contains?(order_tax_zone)
+        rates_for_default_zone << rate if rate.default_vat?
```

Comment:
Whoops, you're right. This way I just run the `contains?` call twice pre iteration, nothing gained.